### PR TITLE
Feature/vpm

### PR DESCRIPTION
**Pull request by zebengberg**
Closes #XX

### Features

- Add a new `pycontrails.models.extended_k15` module for modeling volatile particulate matter (vPM) emissions and activation in aircraft plumes. This module implements the extended K15 model introduced by [Ponsonby et al.](https://doi.org/10.5194/egusphere-2025-1717).
- Introduce basic support for vPM activation within the `Cocip` and `CocipGrid` models. This support is experimental and can be enabled via the `vpm_activation` parameter.

### Internals

- Refactor `contrail_properties.initial_ice_particle_number` to integrate more smoothly with the vPM activation model. The function's signature has been updated, but its underlying behavior is unchanged. In particular, this has no effect on the `Cocip` model when `vpm_activation` is disabled.

## Tests

- [x] QC test passes locally (`make test`)
- [ ] CI tests pass

## Reviewer

@jponvc @roger-teoh

### CODE REVIEW

File: pycontrails/models/extended_k15.py [hunk 1]
```diff
@@ -0,0 +1,1352 @@
...
+    val0 = np.take_along_axis(f, i0, axis=-1)
+    dist = val0 / (val0 - val1)
+
+    # When f never attains a positive value, set i0 and i1 to last negative value
+    # If f never attains a negative value, we pass through a nan
+    cond = attains_negative & ~attains_positive
+    last_negative = np.nanargmax(f[cond], axis=-1, keepdims=True)
+    i0[cond] = last_negative
+    i1[cond] = last_negative
+    dist[cond] = 0.0
```
Comment (zebengberg) [inline pycontrails/models/extended_k15.py:1281]: @roger-teoh @jponvc I did a bunch of exploration in the case when `f` never crosses zero. I do think that by increasing the resolution of the plume sampling, you can always get a zero, but I'm also seeing that it's totally fine to just take the last value of `n_available_all` before the nans start appearing in `n_2_w`. I ran the model on a sample of Roger's GAIA data with 10, 20, 40, 60, 100, 150, 200 test points along the plume. The output was very stable even after 40 test points (I chose a default of 50 here in pycontrails, but it's easily configurable).

| `n_plume_points` | Mean relative error compared to `n_plume_points=200`|
|----------------|------------------|
| 10             | 7.237%           |
| 20             | 2.755%           |
| 40             | 0.711%           |
| 60             | 0.304%           |
| 100            | 0.091%           |
| 150            | 0.026%           |

(It's around here in the code where I overwrite the output in when `f` never becomes positive.)

File: pycontrails/models/extended_k15.py [hunk 2]
```diff
@@ -0,0 +1,1352 @@
...
+import scipy.optimize
+import scipy.special
+import scipy.stats
+
+from pycontrails.physics import constants, thermo
+
+# See upcoming Teoh et. al paper "Impact of Volatile Particulate Matter on Global Contrail
+# Radiative Forcing and Mitigation Assessment" for details on these default parameters.
+DEFAULT_VPM_EI_N = 2.0e17  # vPM number emissions index, [kg^-1]
+DEFAULT_EXHAUST_T = 600.0  # Exhaust temperature, [K]
```
Comment (zebengberg) [inline pycontrails/models/extended_k15.py:23]: @roger-teoh Double check these please!

File: pycontrails/models/extended_k15.py [hunk 3]
```diff
@@ -0,0 +1,1352 @@
...
+    """Define particle types representing nvPM, vPM, and ambient particles.
+
+    See upcoming Teoh et. al paper "Impact of Volatile Particulate Matter on Global Contrail
+    Radiative Forcing and Mitigation Assessment" for details on these default parameters.
+    """
+    return [
+        Particle(type=ParticleType.NVPM, kappa=0.005, gmd=30.0e-9, gsd=2.0, n_ambient=0.0),
+        Particle(type=ParticleType.VPM, kappa=0.2, gmd=1.8e-9, gsd=1.5, n_ambient=0.0),
+        Particle(type=ParticleType.AMBIENT, kappa=0.5, gmd=30.0e-9, gsd=2.3, n_ambient=600.0e6),
+    ]
```
Comment (zebengberg) [inline pycontrails/models/extended_k15.py:85]: @roger-teoh

File: pycontrails/models/extended_k15.py [hunk 4]
```diff
@@ -0,0 +1,1352 @@
...
+    # and temperature. The coefficients below were derived from fitting a linear model to
+    # approximate eps (defined by S(D) = (1 + eps) * Dd) as a function of log(kappa) and log(Dd).
+    # (Dd = 1e-9, 1e-8, 1e-7, 1e-6; kappa = 0.005, 0.05, 0.5; temperature ~= 220 K)
+    b0 = 12.21
+    b_kappa = 0.5883
+    b_Dd = 0.6319
+
+    log_eps = b0 + b_kappa * np.log(kappa) + b_Dd * np.log(Dd)
+    eps = np.exp(log_eps)
+    return (1.0 + eps) * Dd
```
Comment (zebengberg) [inline pycontrails/models/extended_k15.py]: @jponvc

Like I mentioned earlier, I put a lot of thought into an initial guess for Newton's method (to find the maximum of `S(D)`). For a wide range of values of `kappa` and `Dd`, I found that `log(epsilon)` is approximately linear in `log(kappa)` and `log(Dd)`. Here `epsilon` is how you define it: `D = (1 + epsilon) Dd`. (Making a good initial guess massively speeds up Newton's method.)

File: pycontrails/models/extended_k15.py (hunk 1 above)
Comment (jponvc) [reply pycontrails/models/extended_k15.py:1281]: That's good to hear - I think that taking the asymptotic value is physically reasonable.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,16 @@
 # Changelog
 
+## 0.55.0
+
+### Features
+
+- Add a new `pycontrails.models.extended_k15` module for modeling volatile particulate matter (vPM) emissions and activation in aircraft plumes. This module implements the extended K15 model introduced by [Ponsonby et al.](https://doi.org/10.5194/egusphere-2025-1717).
```
Comment (roger-teoh) [inline CHANGELOG.md]: Language correctness: The extended K15 only models vPM activation, not the emissions. The vPM emissions and properties are inputs to the extended K15.

File: pycontrails/models/cocip/contrail_properties.py [hunk 6]
```diff
@@ -202,29 +202,29 @@ def iwc_post_wake_vortex(
...
+
+        The signature of this function has changed. The parameter ``aei`` is now
+        expected to be the previous ``npvm_ei_n`` multiplied by the activation
+        fraction ``f_activation``.
+
     Parameters
     ----------
-    nvpm_ei_n : npt.NDArray[np.floating]
-        black carbon number emissions index, [:math:`kg^{-1}`]
+    aei : npt.NDArray[np.floating]
```
Comment (roger-teoh) [inline pycontrails/models/cocip/contrail_properties.py:224]: aei is the "Apparent contrail emissions index"

File: pycontrails/models/cocip/contrail_properties.py [hunk 7]
```diff
@@ -235,9 +235,8 @@ def initial_ice_particle_number(
         The initial number of ice particles per distance before the wake vortex
         phase, [:math:`# m^{-1}`]
     """
-    f_activation = ice_particle_activation_rate(air_temperature, T_crit_sac)
-    nvpm_ei_n_activated = nvpm_ei_n * f_activation
-    return fuel_dist * np.maximum(nvpm_ei_n_activated, min_ice_particle_number_nvpm_ei_n)
+    aei_clipped = np.clip(aei, min=min_ice_particle_number_nvpm_ei_n)  # type: ignore[arg-type,call-overload]
```
Comment (roger-teoh) [inline pycontrails/models/cocip/contrail_properties.py]: (1) Refactor `min_ice_particle_number_nvpm_ei_n` to `min_aei`,

(2) I understand this following approach is not in pycontrails-vpm, but we should probably only clip `min_aei` if vPM activation is deactivated. This is because the clipping was previously included to floor the potential for vPM and ambient particle activation. This is no longer needed with the extended K15.

File: pycontrails/models/extended_k15.py [hunk 8]
```diff
@@ -0,0 +1,1290 @@
...
+
+    type: ParticleType
+    kappa: float
+    gmd: float
+    gsd: float
+    n_ambient: float
+
+    def __post_init__(self) -> None:
+        if self.type != ParticleType.AMBIENT and self.n_ambient != 0.0:
+            raise ValueError(f"n_ambient must be 0.0 for emitted particle type {self.type.value}")
```
Comment (roger-teoh) [inline pycontrails/models/extended_k15.py]: for "aircraft-emitted" particle type

File: pycontrails/models/extended_k15.py [hunk 9]
```diff
@@ -0,0 +1,1290 @@
...
+    n_available: npt.NDArray[np.floating]
+
+
+def S(
+    D: npt.NDArray[np.floating],
+    Dd: npt.NDArray[np.floating],
+    kappa: npt.NDArray[np.floating],
+    A: npt.NDArray[np.floating],
+) -> npt.NDArray[np.floating]:
+    """Compute the supersaturation ratio at diameter ``D``.
```
Comment (roger-teoh) [inline pycontrails/models/extended_k15.py:119]: Doc strings for input and output variables?

File: pycontrails/models/extended_k15.py [hunk 10]
```diff
@@ -0,0 +1,1290 @@
...
+
+    Implements equation (6) in Petters and Kreidenweis (2007).
```
Comment (roger-teoh) [inline pycontrails/models/extended_k15.py]: Should the citation be included in the biblography?

File: pycontrails/models/extended_k15.py [hunk 11]
```diff
@@ -0,0 +1,1290 @@
...
+    den = (N * (N + c)) ** 2
+
+    return num / den
+
+
+def _newton_seed(
+    Dd: npt.NDArray[np.floating],
+    kappa: npt.NDArray[np.floating],
+) -> npt.NDArray[np.floating]:
+    # This is a crude approach, but it probably works well enough for common values of kappa, Dd,
```
Comment (roger-teoh) [inline pycontrails/models/extended_k15.py]: Add/update doc-string?

File: pycontrails/models/extended_k15.py [hunk 12]
```diff
@@ -0,0 +1,1290 @@
...
+    i_T_ub = np.where(filt, np.arange(T_plume_test.shape[-1]), -1).max(axis=-1, keepdims=True)
+    T_ub = np.take_along_axis(T_plume_test, i_T_ub, axis=-1) + step
+
+    # Now create n_points linearly-spaced values from T_ub down to T_lb
+    # (We assume later that T_plume is sorted in descending order, so we slice [::-1])
+    points = np.linspace(0.0, 1.0, n_points, dtype=float)
+    return (T_lb + (T_ub - T_lb) * points)[..., ::-1]
+
+
+def droplet_apparent_emission_index(
```
Comment (roger-teoh) [inline pycontrails/models/extended_k15.py:407]: Should we move this function to the top (i.e., first function in this script), since it is the main function of `extended_K15.py`?

File: pycontrails/models/extended_k15.py [hunk 13]
```diff
@@ -0,0 +1,1290 @@
...
+    T_ambient: npt.NDArray[np.floating],
+    T_exhaust: npt.NDArray[np.floating],
+    air_pressure: npt.NDArray[np.floating],
+    nvpm_ei_n: npt.NDArray[np.floating],
+    vpm_ei_n: npt.NDArray[np.floating],
+    G: npt.NDArray[np.floating],
+    particles: list[Particle] | None = None,
+    n_plume_points: int = 50,
+) -> npt.NDArray[np.floating]:
+    """Calculate droplet apparent ice emissions index from nvPM, vPM and ambient particles.
```
Comment (roger-teoh) [inline pycontrails/models/extended_k15.py]: Droplet apparent emissions index, not droplet apparent ice emissions index. 

Here, we assume that all the droplets would freeze to form contrail ice crystals.

File: tests/unit/test_cocip.py [hunk 14]
```diff
@@ -1909,6 +1909,39 @@ def test_cocip_survival_fraction(fl: Flight, met: MetDataset, rad: MetDataset):
...
+        met=met,
+        rad=rad,
+        dt_integration="5 minutes",
+        max_age="10 minutes",
+        process_emissions=False,
+        humidity_scaling=ConstantHumidityScaling(rhi_adj=0.6),
+        filter_sac=False,
+        vpm_activation=vpm_activation,
+    )
+    cocip.eval(fl)
```
Comment (roger-teoh) [inline tests/unit/test_cocip.py:1937]: It would be great to overwrite the `nvpm_ei_n` in `fl` to 1e11. By doing so, the unit tests would show a significantly greater difference in `mean_n_ice_per_m_0` and `mean_ef` with and without vPM activation.

File: tests/unit/test_vpm.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+    r_act_finite = r_act[~np.isnan(r_act)]
+    assert np.all(r_act_finite > 0.0)
+
+    # Confirm the min and max are between 1e-9 and 2e-6
+    assert np.min(r_act_finite) > 1e-9
+    assert np.max(r_act_finite) < 2e-7
+
+
+def test_droplet_apparent_emission_index() -> None:
```
Comment (roger-teoh) [inline tests/unit/test_vpm.py:49]: Should we also include tests for the edge cases identified?

File: pycontrails/models/extended_k15.py [hunk 16]
```diff
@@ -0,0 +1,1290 @@
+"""Support for volatile particulate matter (vPM) modeling via the extended K15 model.
+
```
Comment (roger-teoh) [inline pycontrails/models/extended_k15.py:2]: Awaiting for @jponvc to draft up the UserWarning to be shown when the model is initialised

Comment (jponvc) [reply pycontrails/models/extended_k15.py:2]: Thanks for the placeholder @roger-teoh. Please could you use the following text for the UserWarning: "This model is a minimal framework used to approximate the apparent emission index of contrail ice crystals formed in the jet regime (DOI: 10.5194/egusphere-2025-1717). Therefore, it does not fully represent the complexity of microphysical plume processes, including the formation and growth of volatile particulate matter (vPM). Instead, vPM properties are prescribed as model inputs, which strongly impact model outputs. Therefore, the model should only be used for research purposes, together with thorough sensitivity analyses or with explicit reference to the limitations/uncertainties outlined above."

File: CHANGELOG.md (hunk 5 above)
Comment (zebengberg) [reply CHANGELOG.md]: 3b8c8775f54e874ceb03893ada3cf841dcbb39c5

File: pycontrails/models/cocip/contrail_properties.py (hunk 7 above)
Comment (zebengberg) [reply pycontrails/models/cocip/contrail_properties.py]: Makes sense!

a9e588c4915400a32c67c1da0eb00f5a738c3688

File: pycontrails/models/cocip/contrail_properties.py (hunk 6 above)
Comment (zebengberg) [reply pycontrails/models/cocip/contrail_properties.py:224]: a9e588c4915400a32c67c1da0eb00f5a738c3688

File: pycontrails/models/extended_k15.py (hunk 16 above)
Comment (zebengberg) [reply pycontrails/models/extended_k15.py:2]: 17b8f04dad59e3f8e73a3468346e5caee2584957

File: pycontrails/models/extended_k15.py (hunk 13 above)
Comment (zebengberg) [reply pycontrails/models/extended_k15.py]: 17b8f04dad59e3f8e73a3468346e5caee2584957

File: pycontrails/models/extended_k15.py (hunk 11 above)
Comment (zebengberg) [reply pycontrails/models/extended_k15.py]: 17b8f04dad59e3f8e73a3468346e5caee2584957

File: pycontrails/models/extended_k15.py (hunk 9 above)
Comment (zebengberg) [reply pycontrails/models/extended_k15.py:119]: 17b8f04dad59e3f8e73a3468346e5caee2584957

File: tests/unit/test_cocip.py (hunk 14 above)
Comment (zebengberg) [reply tests/unit/test_cocip.py:1937]: 9b924e0273df161b1509c363617bf4b7b0e68c4a

File: pycontrails/models/extended_k15.py (hunk 10 above)
Comment (zebengberg) [reply pycontrails/models/extended_k15.py]: 5fe55ec7a6494069f9a2218bdb339c9b1ff77f16

File: tests/unit/test_vpm.py (hunk 15 above)
Comment (zebengberg) [reply tests/unit/test_vpm.py:49]: I didn't test all of the edge cases that we identified, but I did include a few more tests comparing the pycontrails implementation to [Joel's notebook](https://github.com/jponvc/extended-K15-model/blob/main/extended-K15-model-notebook.ipynb).

8e28d7d7e5c31a40442d1216b9361a834a9da443

File: pycontrails/models/extended_k15.py (hunk 12 above)
Comment (zebengberg) [reply pycontrails/models/extended_k15.py:407]: e3e5c989a8f121eee14423df180ddebedd90db77
